src/main/scala/shoppinglist/domain/ShoppingList.scala

```diff
@@ -0,0 +1,66 @@
+package shoppinglist.domain
+
+import shoppinglist.api.{ListId, ListItemId, OwnerId, ShoppingListManagementError}
+import ShoppingListManagementError._
+import zio.{UIO, ZIO}
+
+final case class ShoppingList(
+    id: ListId,
+    ownerId: OwnerId,
+    name: String,
+    items: List[ListItem]
+) {
+
+  def addItem(itemId: ListItemId, name: String, quantity: Int): Either[ShoppingListManagementError, ShoppingList] =
+    if (items.exists(_.id == itemId))
+      Left(DuplicateItemId)
+    else if (items.exists(_.name == name))
```

Comment:
But you would need to check this constraint on a repository level anyway. If list is shared duplication could happen again.

Reply:
Due to the aggregate being a transaction boundary we gonna use `SELECT FOR UPDATE` and that would lock all the rows until the transaction gets committed. That would prevent possible duplication. 
On the other hand if you're building in a highly collaborative domain I guess then it would make sense to either work with `ListItem` as a root entity and delegate the uniqueness checks to the database, or use something like CRDT.